subprojects/core/src/main/java/org/gradle/initialization/StartParameterBuildOptionFactory.java

```diff
@@ -0,0 +1,308 @@
+        }
+
+        @Override
+        public void applyTo(List<String> values, StartParameter settings) {
+            settings.setExcludedTaskNames(values);
+        }
+    }
+
+    public static class IncludeBuildOption extends ListBuildOption<StartParameter> {
+        public IncludeBuildOption() {
+            super(null, CommandLineOptionConfiguration.create("include-build", "Include the specified build in the composite.").incubating());
+        }
+
+        @Override
+        public void applyTo(List<String> values, StartParameter settings) {
+            Transformer<File, String> resolver = new BasicFileResolver(settings.getCurrentDir());
+
+            for (String includedBuild : values) {
+                settings.includeBuild(resolver.transform(includedBuild));
+            }
+        }
+    }
+
+    public static class ConfigureOnDemandOption extends BooleanBuildOption<StartParameter> {
+        public static final String GRADLE_PROPERTY = "org.gradle.configureondemand";
+
+        public ConfigureOnDemandOption() {
+            super(GRADLE_PROPERTY, CommandLineOptionConfiguration.create("configure-on-demand", "Configure necessary projects only. Gradle will attempt to reduce configuration time for large multi-project builds.").incubating());
+        }
+
+        @Override
+        public void applyTo(boolean value, StartParameter settings) {
+            settings.setConfigureOnDemand(value);
+        }
+    }
+
+    public static class BuildCacheOption extends BooleanBuildOption<StartParameter> {
+        public static final String GRADLE_PROPERTY = "org.gradle.caching";
+
+        public BuildCacheOption() {
+            super(GRADLE_PROPERTY, CommandLineOptionConfiguration.create("build-cache", "Enables the Gradle build cache. Gradle will try to reuse outputs from previous builds.").incubating());
+        }
+
+        @Override
+        public void applyTo(boolean value, StartParameter settings) {
+            settings.setBuildCacheEnabled(value);
+        }
+    }
+
+    public static class BuildScanOption extends BooleanBuildOption<StartParameter> {
+        public BuildScanOption() {
+            super(null, CommandLineOptionConfiguration.create("scan", "Creates a build scan. Gradle will emit a warning if the build scan plugin has not been applied. (https://gradle.com/build-scans)").incubating());
+        }
+
+        @Override
+        public void configure(CommandLineParser parser) {
+            if (hasCommandLineOption()) {
+                String disabledOption = getDisabledCommandLineOption();
+                parser.option(commandLineOptionConfiguration.getLongOption()).hasDescription(commandLineOptionConfiguration.getDescription());
+                parser.option(disabledOption).hasDescription(getDisabledCommandLineDescription());
+                // does not behave like a regular boolean option and therefore doesn't call CommandLineParser.allowOneOf
```

Comment:
It should. The validation below in `applyFromCommandLine()` does exactly what `allowOneOf()` would do. We should get rid of the special case handling of this option.

Reply:
We should check with the CS team before making the change. There might be good reasons why that have that logic. Some test cases in Gradle core actually verify that behavior. Should probably be addressed in a different change.

Reply:
This was [discussed internally](https://gradle.slack.com/archives/C03A0MT6S/p1504206522000222). The main reason for having custom behavior is that build scans can be enabled programmatically via `publishAlways()` which is not the case for other options.

Reply:
This custom logic _does the same thing as_ `allowOneOf()`. It's not _actually_ custom behaviour. The difference is only in how the value is set on the `StartParameter`. There's no difference in how the command-line option behaves from the user's perspective. And the implementation should reflect this.

We should remove all the custom command-line option handling, and instead have `applyTo(v, s)` call `StartParameter.setBuildScan(true)` when `v` is `true` and `StartParameter.setNoBuildScan(true)` when `v` is `false`.

Reply:
I had that at some point until I realized that the test `CommandLineIntegrationSpec.cannot combine --scan and --no-scan` was failing. I think in that case we are OK with removing the test.